### PR TITLE
Use best epoch to choose optimised parameters

### PR DESCRIPTION
**Pull request by ecole41**
This PR edits the `monte_carlo_fit.py` such that the "best epoch" is chosen during training. This is the last epoch from which the validation loss improved with respect to the best validation loss throughout training AND the positivity datasets pass with a threshold of 1e-6. This "best epoch" is then used to produce the outputted optimised parameters

This matches the method used in NNPDF and may be useful for NTK analysis.

To Do:
- [x] Add $\chi^2$ threshold to the best epoch selection
- [x] Add documentation

### CODE REVIEW

File: colibri/doc/sphinx/source/tutorials/scripts/fit-folders.rst
Comment (vschutze-alt) [inline colibri/doc/sphinx/source/tutorials/scripts/fit-folders.rst:1]: Hi Ella, thanks for this update to the documentation. Have you tried to make it locally and see how it looks like once it is compiled? At the moment the "Best epoch" heading doesn't follow the hierarchy in the file and so the "Postfit selection" heading is not being compiled.

**Review by vschutze-alt — CHANGES_REQUESTED** [review]

File: colibri/doc/sphinx/source/tutorials/scripts/fit-folders.rst [hunk 1]
```diff
@@ -119,6 +119,14 @@ the first few lines would look like this:
 
 which would represent the losses for the first 150 epochs (i.e. 0, 1, 2 are just labels).
 
+Best epoch
+-----------
```
Comment (vschutze-alt) [inline colibri/doc/sphinx/source/tutorials/scripts/fit-folders.rst]:
```suggestion
----------
```

**Comment by vschutze-alt** [comment]
report comparing with main: https://vp.nnpdf.science/6NZ5rF7sSd2sb0EtRC8dNQ==
This is a closure test with the les houches model. The uncertainties and central values seem to be improved with the best epoch PR.

**Comment by ecole41** [comment]
> report comparing with main: https://vp.nnpdf.science/6NZ5rF7sSd2sb0EtRC8dNQ== This is a closure test with the les houches model. The uncertainties and central values seem to be improved with the best epoch PR.

That's great, thanks for running this. This would be expected as we are now choosing the optimised parameters throughout the whole of the training of a replica. Out of curiosity, what sort of closure test are these results from? L0, L1 or L2?

**Review by vschutze-alt — APPROVED** [review]

File: colibri/gradient_descent.py [hunk 2]
```diff
@@ -36,6 +36,9 @@ def run_gradient_descent(
     max_epochs: int,
     data_batch: Optional[colibri.DataBatches] = None,
     record_every: int = 50,
+    positivity_check_fn: Optional[Callable[[jnp.ndarray], bool]] = None,
```
Comment (LucaMantani) [inline colibri/gradient_descent.py]: Missing the docstring of this function below.

File: colibri/gradient_descent.py [hunk 3]
```diff
@@ -114,10 +132,35 @@ def _gen():
...
+        epoch_val_chi2 = epoch_val_loss / validation_ndata
+        update_best = False
+        meets_threshold = epoch_val_chi2 < threshold_chi2
+        if meets_threshold:
+            if pos_pass and not any_pos_pass:
+                update_best = True
+                any_pos_pass = True
+            elif pos_pass == any_pos_pass and pos_pass:
+                if epoch_val_loss < best_val_loss:
+                    update_best = True
```
Comment (LucaMantani) [inline colibri/gradient_descent.py]:
```suggestion
        if meets_threshold and pos_pass:
    if not any_pos_pass:
        # first passing epoch: accept it regardless of loss
        update_best = True
        any_pos_pass = True
    elif epoch_val_loss < best_val_loss:
        update_best = True
```
I think it reads easier this way? it should be equivalent if I am not mistaken.

File: colibri/gradient_descent.py [hunk 4]
```diff
@@ -127,13 +170,32 @@ def _gen():
...
+        best_epoch_dict = {
+            "epoch": epoch,
+            "best_parameters": params,
+            "best_val_loss": epoch_val_loss,
+            "best_train_loss": epoch_train_loss,
+        }
+        best_params = params
+    else:
+        best_epoch_dict = {
+            "epoch": best_epoch_idx,
```
Comment (LucaMantani) [inline colibri/gradient_descent.py]: Why here best_parameters is missing in this dictionary?

**Review by LucaMantani — CHANGES_REQUESTED** [review]
I like it overall, I have some minor suggestions to change and a couple of comments / questions:
1) Do you see a slow down of the gradient descent because of the additional things?
2) At the moment the hessian is passing jnp.nan as validation, should we pass the training loss to take advantage of the feature?

File: colibri/gradient_descent.py (hunk 4 above)
Comment (ecole41) [reply colibri/gradient_descent.py]: The dictionary before this `else` should also not have the parameters in the dictionary. It's unnecessary as the parameters are saved regardless in the `optimized_parameters`. I have now edited this file so that this is the case

File: colibri/gradient_descent.py [hunk 5]
```diff
@@ -36,6 +36,9 @@ def run_gradient_descent(
     max_epochs: int,
     data_batch: Optional[colibri.DataBatches] = None,
     record_every: int = 50,
+    positivity_check_fn: Optional[Callable[[jnp.ndarray], bool]] = None,
+    threshold_chi2: float = 10.0,
+    validation_ndata: int = 1,  # If not defined then original loss will be used - As in Hessian
```
Comment (LucaMantani) [inline colibri/gradient_descent.py]: This can be removed completely now, the validation_loss_fn has ndata as an attribute.

**Review by LucaMantani — COMMENTED** [review]
~There is no need anymore to pass the validation_ndata now, you can directly access it from the function now.~

Sorry, I actually overlooked the fact that we are not passing the Likelihood objects but we are defining simple functions. What I was thinking can only be done if we change a bit the code and pass directly the Loglikelihood objects. I think it would probably be nice to do, but in a separate PR that does only that specific thing.
